### PR TITLE
more useful graphite errors

### PR DESCRIPTION
**Pull request by Dieterbe**
* make errors clearer in the web ui (shows request url and more information about what went wrong)
* print more details to log. in particular, the response body with graphite stacktrace:

```
2015/04/08 18:10:12 bosun web listening on: :8070
2015/04/08 18:10:12 tsdb host: 
2015/04/08 18:10:18 error: funcs.go:351: graphite: Get(http://graphite/render/?format=json&from=1428271818&target=divideSeries%28stats.gauges.stats_buffer_space_mb.%2A._data_s3%2A%29&until=1428531018) failed: got statuscode 500 INTERNAL SERVER ERROR
===================================================
<body style="background-color: #666666; color: black;">
<center>
<h2 style='font-family: "Arial"'>
<p>Graphite encountered an unexpected error while handling your request.</p>
<p>Please contact your site administrator if the problem persists.</p>
</h2>
<br/>
<div style="width: 50%; text-align: center; font-family: monospace; background-color: black; font-weight: bold; color: #ff4422;">

</div>

<div style="width: 70%; text-align: left; background-color: black; color: #44ff22; border: thin solid gray;">
<pre>
Traceback (most recent call last):
  File &quot;/usr/lib/python2.6/site-packages/django/core/handlers/base.py&quot;, line 109, in get_response
    response = callback(request, *callback_args, **callback_kwargs)
  File &quot;/usr/lib/python2.6/site-packages/graphite/render/views.py&quot;, line 107, in renderView
    seriesList = evaluateTarget(requestContext, target)
  File &quot;/usr/lib/python2.6/site-packages/graphite/render/evaluator.py&quot;, line 10, in evaluateTarget
    result = evaluateTokens(requestContext, tokens)
  File &quot;/usr/lib/python2.6/site-packages/graphite/render/evaluator.py&quot;, line 21, in evaluateTokens
    return evaluateTokens(requestContext, tokens.expression)
  File &quot;/usr/lib/python2.6/site-packages/graphite/render/evaluator.py&quot;, line 29, in evaluateTokens
    return func(requestContext, *args)
TypeError: divideSeries() takes exactly 3 arguments (2 given)

</pre>
</div>

</center>

===================================================
2015/04/08 18:10:18 graphite: Get(http://graphite/render/?format=json&from=1428271818&target=divideSeries%28stats.gauges.stats_buffer_space_mb.%2A._data_s3%2A%29&until=1428531018) failed: got statuscode 500 INTERNAL SERVER ERROR
2015/04/08 18:10:22 wrote metrictags: 86.00B
2015/04/08 18:10:22 wrote notifications: 48.00B
```
note:
* the error actually gets logged twice. we should probably fix this, not sure what's the best approach
* it seems overkill to use graphite.Error for those errors where there's no body to show. i'm planning to make those regular errors again.
* couldn't find an elegant way to strip html tags out. (see http://grokbase.com/t/gg/golang-nuts/13ckt90wpg/go-nuts-export-striptags-in-html-template). I could maybe try https://godoc.org/golang.org/x/net/html but that looks quite complicated too

### CODE REVIEW

File: graphite/graphite.go [hunk 1]
```diff
@@ -26,12 +26,30 @@ type Series struct {
 
 type DataPoint []json.Number
 
+type Error struct {
```
Comment (mjibson) [inline graphite/graphite.go]: I'd prefer something like:

```
type Error struct {
  URL url.URL
  StatusCode int
  Body []byte // response body
}
```

All the fields are exportable and thus usable from timeGraphiteRequest.

File: graphite/graphite.go [hunk 2]
```diff
@@ -26,12 +26,30 @@ type Series struct {
...
+	subject string
+	body    string
+	err     error
+}
+
+func (e Error) Error() string {
+	return fmt.Sprintf("graphite: %s: %s", e.subject, e.err.Error())
+}
+
+func (e Error) Body() string {
```
Comment (mjibson) [inline graphite/graphite.go]: Who are we? Java?

File: graphite/graphite.go [hunk 3]
```diff
@@ -68,13 +108,13 @@ var DefaultClient = &http.Client{
 
 // Context is the interface for querying a Graphite server.
 type Context interface {
-	Query(*Request) (Response, error)
+	Query(*Request) (Response, *Error)
```
Comment (mjibson) [inline graphite/graphite.go]: This should still be error.

File: graphite/graphite.go [hunk 4]
```diff
@@ -68,13 +108,13 @@ var DefaultClient = &http.Client{
...
 }
 
 // Host is a simple Graphite Context with no additional features.
 type Host string
 
 // Query performs a request to a Graphite server.
-func (h Host) Query(r *Request) (Response, error) {
+func (h Host) Query(r *Request) (Response, *Error) {
```
Comment (mjibson) [inline graphite/graphite.go]: error here still

File: graphite/graphite.go [hunk 5]
```diff
@@ -42,23 +60,45 @@ func (r *Request) Query(host string) (Response, error) {
...
 		RawQuery: v.Encode(),
 	}
-	resp, err := DefaultClient.Get(u.String())
+}
+
+func (r *Request) String() string {
+	return r.url.String()
+}
+
+func (r *Request) Query(host string) (Response, *Error) {
```
Comment (mjibson) [inline graphite/graphite.go]: just error

File: graphite/graphite.go [hunk 6]
```diff
@@ -42,23 +60,45 @@ func (r *Request) Query(host string) (Response, error) {
...
+	r.url = url.URL{
 		Scheme:   "http",
 		Host:     host,
 		Path:     "/render/",
 		RawQuery: v.Encode(),
 	}
-	resp, err := DefaultClient.Get(u.String())
+}
+
+func (r *Request) String() string {
```
Comment (mjibson) [inline graphite/graphite.go]: shouldn't need this method

File: cmd/bosun/expr/funcs.go [hunk 7]
```diff
@@ -355,7 +364,7 @@ func parseGraphiteResponse(req *graphite.Request, s *graphite.Response, formatTa
 		// build tag set
 		nodes := strings.Split(res.Target, ".")
 		if len(nodes) < len(formatTags) {
-			return nil, fmt.Errorf(`returned target "%s" does not match format "%s"`, res.Target, formatTags)
+			return nil, graphite.NewError(req.String(), "", fmt.Errorf(`returned target "%s" does not match format "%s"`, res.Target, formatTags))
```
Comment (mjibson) [inline cmd/bosun/expr/funcs.go]: parseGraphiteResponse should not return a graphite.Error because those are limited to web errors, not parse.

File: graphite/graphite.go (hunk 1 above)
Comment (mjibson) [inline graphite/graphite.go]: And maybe rename this to RequestError. Error is too vague.

File: cmd/bosun/expr/funcs.go [hunk 8]
```diff
@@ -344,6 +345,14 @@ func DropNA(e *State, T miniprofiler.Timer, series *Results) (*Results, error) {
 	}
 	return DropValues(e, T, series, 0, dropFunction)
 }
+func printGraphiteError(err error) {
```
Comment (mjibson) [inline cmd/bosun/expr/funcs.go]: This function can be removed since there's only one place this kind of error can be produced, so just inline it.

File: cmd/bosun/expr/funcs.go [hunk 9]
```diff
@@ -344,6 +345,14 @@ func DropNA(e *State, T miniprofiler.Timer, series *Results) (*Results, error) {
...
+	if _, ok := err.(*graphite.Error); ok {
```
Comment (mjibson) [inline cmd/bosun/expr/funcs.go]: if err, ok := ..., then you don't need to do the assertion below

File: graphite/graphite.go [hunk 10]
```diff
@@ -26,12 +26,30 @@ type Series struct {
...
+	return &Error{subject, body, err}
+}
+
 func (r *Request) CacheKey() string {
 	targets, _ := json.Marshal(r.Targets)
 	return fmt.Sprintf("graphite-%d-%d-%s", r.Start.Unix(), r.End.Unix(), targets)
 }
 
-func (r *Request) Query(host string) (Response, error) {
+func (r *Request) setUrl(host string) {
```
Comment (mjibson) [inline graphite/graphite.go]: Why does this need to be its own function?

File: graphite/graphite.go [hunk 11]
```diff
@@ -42,23 +60,45 @@ func (r *Request) Query(host string) (Response, error) {
...
+	resp, err := DefaultClient.Get(r.url.String())
 	if err != nil {
-		return Response{}, err
+		e := NewError(fmt.Sprintf("Get(%s) failed", r.url.String()), "", err)
+		return Response{}, e
 	}
 	defer resp.Body.Close()
 	if resp.StatusCode != http.StatusOK {
-		return Response{}, errors.New(resp.Status)
+		body := make([]byte, 2000)
```
Comment (mjibson) [inline graphite/graphite.go]: ioutil.ReadAll(resp.Body) will do

File: graphite/graphite.go [hunk 12]
```diff
@@ -42,23 +60,45 @@ func (r *Request) Query(host string) (Response, error) {
...
+		var msg string
+		_, err := resp.Body.Read(body)
+		if err == nil {
```
Comment (mjibson) [inline graphite/graphite.go]: if err != nil { return err }

File: graphite/graphite.go (hunk 5 above)
Comment (mjibson) [inline graphite/graphite.go]: and this should really return a *Response so you can return nil

File: graphite/graphite.go (hunk 1 above)
Comment (Dieterbe) [inline graphite/graphite.go]: graphite.Error can conceptually contain error information about anything wrong when doing anything graphite related. it'll probably be mostly/all request related but i wouldn't restrict ourselves if we don't need to.

File: graphite/graphite.go (hunk 2 above)
Comment (Dieterbe) [inline graphite/graphite.go]: i like the idea of having private fields that can only be set at construction time, so we can guarantee they won't accidentally be modified and incorrect. Lemme know if you like this idea or still like to just export the fields directly.

File: graphite/graphite.go (hunk 11 above)
Comment (Dieterbe) [inline graphite/graphite.go]: i wanted to have an upper limit on the amount of data going into the error body, cause the error body will be printed to stdout/log

File: graphite/graphite.go (hunk 6 above)
Comment (Dieterbe) [inline graphite/graphite.go]: what do you mean? i call this in a few places when creating errors to display the request info to the user.

File: cmd/bosun/expr/funcs.go (hunk 8 above)
Comment (Dieterbe) [inline cmd/bosun/expr/funcs.go]: what do you mean, graphite errors are being produced in a variety of places.
And most of those locations are in the graphite library so it wouldn't be good to print the error there.
where do you suggest i inline this and print errors in a central place?

File: graphite/graphite.go (hunk 10 above)
Comment (Dieterbe) [inline graphite/graphite.go]: i find it cleaner, especially since url is now a field of req. the Query() function had too much stuff in it

File: cmd/bosun/expr/funcs.go (hunk 7 above)
Comment (Dieterbe) [inline cmd/bosun/expr/funcs.go]: i want to make this back to a regular error but for the reason that if there's no body (==extra info for log) than there's no need to use graphite.Error.
I do disagree that graphite.Error should only be used for web errors. they are designed to be used any time you want to print extra stuff to the log.

File: graphite/graphite.go (hunk 12 above)
Comment (Dieterbe) [inline graphite/graphite.go]: that would mask the original error of the statuscode being bad. if reading body fails, i want to mention that as additional information, but we care about both errors, not just the last one.

File: cmd/bosun/expr/funcs.go (hunk 9 above)
Comment (Dieterbe) [inline cmd/bosun/expr/funcs.go]: neat. will fix

File: graphite/graphite.go (hunk 5 above)
Comment (Dieterbe) [inline graphite/graphite.go]: right right

File: graphite/graphite.go (hunk 11 above)
Comment (Dieterbe) [inline graphite/graphite.go]: actually i'll do as you say, and do limiting in the "UI" layer where applicable

File: graphite/graphite.go (hunk 1 above)
Comment (Dieterbe) [inline graphite/graphite.go]: actually thinking about it more, we shouldn't have 1 general purpose error clas. will rename to RequestErr and if we ever need other classes we can make em.

File: graphite/graphite.go (hunk 2 above)
Comment (mjibson) [inline graphite/graphite.go]: Just export directly.

File: cmd/bosun/expr/funcs.go (hunk 7 above)
Comment (mjibson) [inline cmd/bosun/expr/funcs.go]: If you want verbose and not verbose errors, they should be errors that specify what the error was, with optional extra data. So if you want errors from different functions, each will need its own type.

**Comment by Dieterbe** [comment]
pushed a bunch of cleanups as discussed and also made it look nice in the web ui:

parse error on expr page:
requesterror-rule

request error on rule page:
graphiteparseerror-expr

i left the dashboard untouched for now, would like to make this look nice too but not sure what i can do
requesterr-dashboard

File: cmd/bosun/expr/funcs.go [hunk 13]
```diff
@@ -657,7 +673,11 @@ func timeGraphiteRequest(e *State, T miniprofiler.Timer, req *graphite.Request)
 		}
 		var val interface{}
 		val, err = e.cache.Get(key, getFn)
-		resp = val.(graphite.Response)
+		// hand out copy of the value, so callers can't accidentally modify data in cache
+		if val, ok := val.(*graphite.Response); ok && val != nil {
```
Comment (mjibson) [inline cmd/bosun/expr/funcs.go]: This won't actually do anything since Response is just a slice.

File: cmd/bosun/web/static/partials/rule.html [hunk 14]
```diff
@@ -95,7 +95,7 @@
 </div>
 <div class="row" ng-show="error">
 	<div class="col-lg-12">
-		<div class="alert alert-danger" ng-bind="error"></div>
+		<pre class="alert alert-danger" ng-bind="error"></pre>
```
Comment (mjibson) [inline cmd/bosun/web/static/partials/rule.html]: This would stop line breaks for long lines. May need some CSS if we decide to stick with the pre.

File: cmd/bosun/web/expr.go [hunk 15]
```diff
@@ -66,6 +67,22 @@ func Expr(t miniprofiler.Timer, w http.ResponseWriter, r *http.Request) (interfa
 	return ret, nil
 }
 
+// presentError formats the error in a way appropriate for a web user to consume
+func presentError(err error) string {
+	switch err := err.(type) {
+	case *graphite.RequestError:
```
Comment (mjibson) [inline cmd/bosun/web/expr.go]: I think it's wrong to have this package depend on the graphite package. Need another solution.

File: cmd/bosun/expr/funcs.go [hunk 16]
```diff
@@ -355,7 +367,7 @@ func parseGraphiteResponse(req *graphite.Request, s *graphite.Response, formatTa
 		// build tag set
 		nodes := strings.Split(res.Target, ".")
 		if len(nodes) < len(formatTags) {
-			return nil, fmt.Errorf(`returned target "%s" does not match format "%s"`, res.Target, formatTags)
+			return nil, graphite.NewParseError(fmt.Sprintf("returned target '%s' does not match format '%s'", res.Target, strings.Join(formatTags, ",")), nil, req.Url)
```
Comment (mjibson) [inline cmd/bosun/expr/funcs.go]: I cannot see any reason in any uses of NewParseError why we wouldn't just use fmt.Errorf instead. The purpose here was to have varying levels of display in the log and the dashboard. I don't think this is needed for that.

Comment (Dieterbe) [inline cmd/bosun/expr/funcs.go]: this tracks the request url as a separate attribute. if we were to put everything together than the error message itself would be too long, we wouldn't have the option of [not] showing the url, and we would have to decide up front how the url should be shown (in dashboard and log we do it in a new line, but i don't think we should formatting such as newlines into the error message)

File: cmd/bosun/web/static/partials/rule.html (hunk 14 above)
Comment (Dieterbe) [inline cmd/bosun/web/static/partials/rule.html]: true. the alternative is that we add `<br>` at every newline, i suppose. but afaik pre is always used for this sort of thing

File: cmd/bosun/expr/funcs.go (hunk 13 above)
Comment (Dieterbe) [inline cmd/bosun/expr/funcs.go]: didn't we have that whole discussion where we concluded processing functions such as avg() can modify result sets in place and so we need to protect by always copying out of the cache?

Comment (mjibson) [inline cmd/bosun/expr/funcs.go]: Yes, but that doesn't happen with graphite data because it is entirely copied later on.

Comment (mjibson) [inline cmd/bosun/expr/funcs.go]: See: https://github.com/bosun-monitor/bosun/blob/2ef959ada8be2e046216aa4e97a72a5b2d6f7832/cmd/bosun/expr/funcs.go#L459

Comment (Dieterbe) [inline cmd/bosun/expr/funcs.go]: how about https://github.com/bosun-monitor/bosun/blob/2ef959ada8be2e046216aa4e97a72a5b2d6f7832/cmd/bosun/expr/funcs.go#L441 (first step of graphiteBand) and https://github.com/bosun-monitor/bosun/blob/2ef959ada8be2e046216aa4e97a72a5b2d6f7832/cmd/bosun/expr/funcs.go#L577-581 (graphiteQuery)

I think the true reason this hasn't been an issue is because we call `parseGraphiteResponse` after timeGraphiteRequest, and this function creates a new Result from scratch each time.

That said, you're right that copying the slice won't be helpful, and that we don't need copying because this is already taken care of elsewhere.

**Comment by Dieterbe** [comment]
opted for simpler approach.
also, systemd journalctl output looks like:
```
Apr 10 16:58:11 dieter-t420s bosun[16752]: 2015/04/10 16:58:11 check alert s3_buffer_bad start
Apr 10 16:58:11 dieter-t420s bosun[16752]: 2015/04/10 16:58:11 graphite RequestError: Get failed: 500 INTERNAL SERVER ERROR
                                           URL: http://graphite/render/?format=json&from=1428440290&target=divideSeries%28stats.gauges.stats_buffer_space_mb.%2A._data_s3%2A%29&until=1428699490
                                           
                                           Traceback:
                                           Traceback (most recent call last):
                                             File &quot;/usr/lib/python2.6/site-packages/django/core/handlers/base.py&quot;, line 109, in get_response
                                               response = callback(request, *callback_args, **callback_kwargs)
                                             File &quot;/usr/lib/python2.6/site-packages/graphite/render/views.py&quot;, line 107, in renderView
                                               seriesList = evaluateTarget(requestContext, target)
                                             File &quot;/usr/lib/python2.6/site-packages/graphite/render/evaluator.py&quot;, line 10, in evaluateTarget
                                               result = evaluateTokens(requestContext, tokens)
                                             File &quot;/usr/lib/python2.6/site-packages/graphite/render/evaluator.py&quot;, line 21, in evaluateTokens
                                               return evaluateTokens(requestContext, tokens.expression)
                                             File &quot;/usr/lib/python2.6/site-packages/graphite/render/evaluator.py&quot;, line 29, in evaluateTokens
                                               return func(requestContext, *args)
                                           TypeError: divideSeries() takes exactly 3 arguments (2 given)
Apr 10 16:58:11 dieter-t420s bosun[16752]: 2015/04/10 16:58:11 check alert s3_buffer_bad done (503.478437ms): 0 crits, 0 warns, 0 unevaluated, 0 unknown
Apr 10 16:58:11 dieter-t420s bosun[16752]: 2015/04/10 16:58:11 s3_buffer_bad{}: template: s3_buffer:1:36: executing "s3_buffer" at <.Eval>: error calling Eval: no results returned
Apr 10 16:58:11 dieter-t420s bosun[16752]: 2015/04/10 16:58:11 check took 990.112214ms
Apr 10 16:58:15 dieter-t420s bosun[16752]: 2015/04/10 16:58:15 wrote metadata: 2.68KB
Apr 10 16:58:15 dieter-t420s bosun[16752]: 2015/04/10 16:58:15 wrote metric: 68.00B
Apr 10 16:58:15 dieter-t420s bosun[16752]: 2015/04/10 16:58:15 wrote tagk: 42.00B
Apr 10 16:58:15 dieter-t420s bosun[16752]: 2015/04/10 16:58:15 wrote tagv: 68.00B
Apr 10 16:58:15 dieter-t420s bosun[16752]: 2015/04/10 16:58:15 wrote metrictags: 86.00B
Apr 10 16:58:15 dieter-t420s bosun[16752]: 2015/04/10 16:58:15 wrote notifications: 54.00B
Apr 10 16:58:15 dieter-t420s bosun[16752]: 2015/04/10 16:58:15 wrote silence: 125.00B
Apr 10 16:58:15 dieter-t420s bosun[16752]: 2015/04/10 16:58:15 wrote status: 14.18KB
Apr 10 16:58:15 dieter-t420s bosun[16752]: 2015/04/10 16:58:15 save to db complete


Apr 10 16:58:25 dieter-t420s bosun[16752]: 2015/04/10 16:58:25 graphite RequestError: Get failed: 500 INTERNAL SERVER ERROR
                                           URL: http://graphite/render/?format=json&from=1428440305&target=divideSeries%28stats.gauges.stats_buffer_space_mb.%2A._data_s3%2A%29&until=1428699505
                                           
                                           Traceback:
                                           Traceback (most recent call last):
                                             File &quot;/usr/lib/python2.6/site-packages/django/core/handlers/base.py&quot;, line 109, in get_response
                                               response = callback(request, *callback_args, **callback_kwargs)
                                             File &quot;/usr/lib/python2.6/site-packages/graphite/render/views.py&quot;, line 107, in renderView
                                               seriesList = evaluateTarget(requestContext, target)
                                             File &quot;/usr/lib/python2.6/site-packages/graphite/render/evaluator.py&quot;, line 10, in evaluateTarget
                                               result = evaluateTokens(requestContext, tokens)
                                             File &quot;/usr/lib/python2.6/site-packages/graphite/render/evaluator.py&quot;, line 21, in evaluateTokens
                                               return evaluateTokens(requestContext, tokens.expression)
                                             File &quot;/usr/lib/python2.6/site-packages/graphite/render/evaluator.py&quot;, line 29, in evaluateTokens
                                               return func(requestContext, *args)
                                           TypeError: divideSeries() takes exactly 3 arguments (2 given)
Apr 10 16:58:35 dieter-t420s slim[1484]: 10-04-2015 16:58:35.444 DEBUG void CommonSystem::monitorNetwork() restarting network - seems to be down!
Apr 10 16:58:35 dieter-t420s slim[1484]: 10-04-2015 16:58:35.444 DEBUG Stopping PcapNetworkTrafficMonitor..
```

File: graphite/graphite.go [hunk 17]
```diff
@@ -68,13 +76,13 @@ var DefaultClient = &http.Client{
 
 // Context is the interface for querying a Graphite server.
 type Context interface {
-	Query(*Request) (Response, error)
```
Comment (mjibson) [inline graphite/graphite.go]: Let's make this change in a separate PR

File: graphite/error.go [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+func NewParseError(message string, err error, url url.URL) error {
```
Comment (mjibson) [inline graphite/error.go]: I do not think either of these methods are needed and will have to be really convinced before I ok them. This one especially is so simple that it does not warrant existing.

File: graphite/error.go [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+	if err != nil {
+		message += ": " + err.Error()
+	}
+	return fmt.Errorf("graphite ParseError: %s\nURL: %s\n", message, url.String())
+}
+
+func NewRequestError(message string, err error, url url.URL, resp *http.Response) error {
+	if err != nil {
+		message += ": " + err.Error()
+	}
```
Comment (mjibson) [inline graphite/error.go]: What is so wrong with just using the whole response body instead of all of this?

Comment (Dieterbe) [inline graphite/error.go]: the whole body is an entire html document with body tags, a h2, some divs that override colors, etc.

File: graphite/graphite.go [hunk 20]
```diff
@@ -15,6 +14,7 @@ type Request struct {
 	Start   *time.Time
 	End     *time.Time
 	Targets []string
+	Url     url.URL
```
Comment (mjibson) [inline graphite/graphite.go]: `Url` -> `URL`: https://github.com/golang/go/wiki/CodeReviewComments#initialisms

File: graphite/error.go [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+		if bodyErr == nil {
+			bodyLines := strings.Split(strings.TrimSpace(string(bodyBytes)), "\n")
+			tracebackLines := make([]string, 0, len(bodyLines))
+			inTraceback := false
+			for _, line := range bodyLines {
+				if strings.HasPrefix(line, "Traceback") {
+					inTraceback = true
+				}
+				if inTraceback && line == "" {
+					inTraceback = false
```
Comment (mjibson) [inline graphite/error.go]: Should also add `if inTraceback { break }` here to stop working

File: graphite/error.go [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+	if resp != nil {
+		bodyBytes, bodyErr := ioutil.ReadAll(resp.Body)
+		if bodyErr == nil {
+			bodyLines := strings.Split(strings.TrimSpace(string(bodyBytes)), "\n")
+			tracebackLines := make([]string, 0, len(bodyLines))
+			inTraceback := false
+			for _, line := range bodyLines {
+				if strings.HasPrefix(line, "Traceback") {
+					inTraceback = true
+				}
```
Comment (mjibson) [inline graphite/error.go]: else if

File: graphite/error.go [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+func NewParseError(message string, err error, url url.URL) error {
+	if err != nil {
+		message += ": " + err.Error()
+	}
+	return fmt.Errorf("graphite ParseError: %s\nURL: %s\n", message, url.String())
+}
+
+func NewRequestError(message string, err error, url url.URL, resp *http.Response) error {
```
Comment (mjibson) [inline graphite/error.go]: The trackback extraction is only used in one place. How about make this function do only that, and then have some error consts to do the rest?

File: graphite/graphite.go (hunk 10 above)
Comment (mjibson) [inline graphite/graphite.go]: Url -> URL

File: graphite/error.go (hunk 23 above)
Comment (Dieterbe) [inline graphite/error.go]: k, sounds good

File: graphite/graphite.go [hunk 24]
```diff
@@ -42,23 +42,31 @@ func (r *Request) Query(host string) (Response, error) {
...
 		Scheme:   "http",
 		Host:     host,
 		Path:     "/render/",
 		RawQuery: v.Encode(),
 	}
-	resp, err := DefaultClient.Get(u.String())
+}
+
+func (r *Request) Query(host string) (Response, error) {
+	r.setUrl(host)
```
Comment (mjibson) [inline graphite/graphite.go]: Why does setUrl need to be its own function? It's only called in one place, and short.

Comment (Dieterbe) [inline graphite/graphite.go]: https://github.com/bosun-monitor/bosun/pull/879#discussion_r28021500

File: graphite/graphite.go [hunk 25]
```diff
@@ -42,23 +47,55 @@ func (r *Request) Query(host string) (Response, error) {
...
 		Scheme:   "http",
 		Host:     host,
 		Path:     "/render/",
 		RawQuery: v.Encode(),
 	}
-	resp, err := DefaultClient.Get(u.String())
+	resp, err := DefaultClient.Get(r.URL.String())
 	if err != nil {
-		return Response{}, err
+		return nil, fmt.Errorf(requestErrFmt, r.URL.String(), "Get failed: "+err.Error())
```
Comment (mjibson) [inline graphite/graphite.go]: just `r.URL` is sufficient, it knows to convert it to a string

File: graphite/graphite.go [hunk 26]
```diff
@@ -3,18 +3,23 @@ package graphite // import "bosun.org/graphite"
...
-	"errors"
 	"fmt"
+	"io/ioutil"
 	"net/http"
 	"net/url"
+	"strings"
 	"time"
 )
 
+const ParseErrFmt = "graphite ParseError (%s): %s"
```
Comment (mjibson) [inline graphite/graphite.go]: Move this to funcs.go if that's the only place it's used

File: graphite/graphite.go [hunk 27]
```diff
@@ -42,23 +47,55 @@ func (r *Request) Query(host string) (Response, error) {
...
+		if err != nil {
+			tb = &[]string{"<Could not read traceback: " + err.Error() + ">"}
+		}
+		return nil, fmt.Errorf(requestErrFmt, r.URL.String(), fmt.Sprintf("Get failed: %s\n%s", resp.Status, strings.Join(*tb, "\n")))
 	}
 	var series Response
 	err = json.NewDecoder(resp.Body).Decode(&series)
-	return series, err
+	if err != nil {
+		e := fmt.Errorf(requestErrFmt, r.URL.String(), "Parsing response failed: "+err.Error())
```
Comment (mjibson) [inline graphite/graphite.go]: since we have 2 kinds of parsing (json and graphite), i'd suggest using "json decode failed" here

File: graphite/graphite.go [hunk 28]
```diff
@@ -42,23 +47,55 @@ func (r *Request) Query(host string) (Response, error) {
...
+	return series, nil
+}
+
+func readTraceback(resp *http.Response) (*[]string, error) {
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	bodyLines := strings.Split(strings.TrimSpace(string(bodyBytes)), "\n")
+	tracebackLines := make([]string, 0, len(bodyLines))
```
Comment (mjibson) [inline graphite/graphite.go]: Let's leave out the cap here. I don't think it's winning us anything. just `var tracebackLines []string` should be enough.

File: graphite/graphite.go (hunk 25 above)
Comment (Dieterbe) [inline graphite/graphite.go]:
```
with String():
http://graphite/render/?format=json&from=1428778698&target=divideSeries%28stats.gauges.stats_buffer_space_mb.%2A._data_s3%2A%29&until=1429037898`

without:
{http  <nil> graphite /render/ format=json&from=1428778588&target=divideSeries%28stats.gauges.stats_buffer_space_mb.%2A._data_s3%2A%29&until=1429037788 }
```
I find the `String()` version clearer, better and also allows direct copy pasting.

File: graphite/graphite.go (hunk 26 above)
Comment (Dieterbe) [inline graphite/graphite.go]: really? to me this makes sense to keep it in the graphite package. semantically (it defines stuff that is specific to graphite) and also because we define the other error here.

but if you feel strongly about this, i guess i'll move it...

Comment (mjibson) [inline graphite/graphite.go]: I feel strongly.

File: graphite/graphite.go (hunk 25 above)
Comment (mjibson) [inline graphite/graphite.go]: Add a `&` to the definition above so it's a `*url.URL` and it'll work correctly.

**Comment by Dieterbe** [comment]
allright. all fixed and squashed together into 1 simple commit.
